### PR TITLE
Enabled drag and drop for Bento.

### PR DESCRIPTION
**Pull request by wyattjsmith1**
This is a preliminary commit for drag and drop in Bento. I am just looking to see if I am headed in a good direction for architecture, then I will flush out all of the details and make it production ready.

For testing, you can run the sample app and click on `Reorder Items`. The page is ugly, but works to demo the feature. There are three sections. One that can not be moved (normal). One that can be rearranged by long pressing and then dragging. Lastly, there is a particularly ugly section for dragging by the handle.

One known limitation of this is that you can long press outside of a handle and still move an item with a handle. I can't seem to find a way to avoid this, but I don't think it's a big deal because you can reorder it anyways.

Another thing to consider: we don't want users to move items to anywhere on the page, so items can only be moved within their component. That means you can not drag an item with a handle to the group with the long press. This means that developers will be required to encompass their entire rearrangeable list within one `Component`. I think that is fine, but I am curious if there are other opinions on this.

### CODE REVIEW

File: bento/src/main/java/com/yelp/android/bento/components/ListComponent.java [hunk 1]
```diff
@@ -253,6 +255,21 @@ public int getSpanSize(int position) {
...
     }
 
+    @Override
+    public void onItemsMoved(int oldIndex, int newIndex) {
+        super.onItemsMoved(oldIndex, newIndex);
+
+        mData.add(newIndex, mData.remove(oldIndex));
+
+        if (mOnItemMovedCallback != null) {
+            mOnItemMovedCallback.onItemMoved(oldIndex, newIndex, mData);
```
Comment (wyattjsmith1) [inline bento/src/main/java/com/yelp/android/bento/components/ListComponent.java]: @dwaxemberg Is it ok to be passing back a raw `mData`, or should I be passing a copy of the list?

File: bento-sample-app/src/main/java/com/yelp/android/bentosampleapp/ReorderListActivity.kt [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        componentController.addComponent(list)
+    }
+
+    private fun addLongPressOrderableListComponent() {
+        componentController.addComponent(LabeledComponent("Long press to reorder"))
+        val longPressComponent =
+                ListComponent<Unit, String>(Unit, LabeledComponentViewHolder::class.java, 2)
+        longPressComponent.setIsReorderable(true)
+        longPressComponent.toggleDivider(false)
+        longPressComponent.setData((0..10).map { 'a'.plus(it).toString() })
```
Comment (dwaxemberg) [inline bento-sample-app/src/main/java/com/yelp/android/bentosampleapp/ReorderListActivity.kt:51]: This would be cleaner with string interpolation: "a $it". Also, why "a"? Can it be something like "Item $it"?

File: bento-sample-app/src/main/java/com/yelp/android/bentosampleapp/ReorderListActivity.kt [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    fun onStartDrag(position: Int)
+}
+
+class ReorderViewHolder : ComponentViewHolder<Presenter, String>() {
+
+    private lateinit var text: TextView
+    private lateinit var presenter: Presenter
+
+    override fun inflate(parent: ViewGroup): View {
+        return LayoutInflater.from(parent.context)
```
Comment (dwaxemberg) [inline bento-sample-app/src/main/java/com/yelp/android/bentosampleapp/ReorderListActivity.kt]: Can be replaced with `parent.inflate()`

File: bento/src/main/java/com/yelp/android/bento/components/ListComponent.java (hunk 1 above)
Comment (dwaxemberg) [reply bento/src/main/java/com/yelp/android/bento/components/ListComponent.java]: Generally it's bad practice unless the list is immutable. You could use Collections.immutableList or convert this to Kotlin and use the immutable List. 

Aside: Why are we passing the list back at all?

File: bento/src/main/java/com/yelp/android/bento/components/ListComponent.java [hunk 4]
```diff
@@ -253,12 +255,42 @@ public int getSpanSize(int position) {
...
+        }
+    }
+
+    public void setOnItemMovedCallback(OnItemMovedCallback<T> callback) {
```
Comment (dwaxemberg) [inline bento/src/main/java/com/yelp/android/bento/components/ListComponent.java:271]: Move the public methods above the private/protected ones

File: bento/src/main/java/com/yelp/android/bento/core/Component.java [hunk 5]
```diff
@@ -262,6 +256,20 @@ final Object getPresenterInternal(int position) {
...
+     * during {@link OnTouchListener#onTouch(android.view.View, android.view.MotionEvent)} if
+     * {@link android.view.MotionEvent} is {@link MotionEvent#ACTION_DOWN}.
+     *
+     * @see Component#onItemsMoved(int, int)
+     */
+    public boolean isReorderable() {
+        return false;
+    }
+
+
```
Comment (dwaxemberg) [inline bento/src/main/java/com/yelp/android/bento/core/Component.java]: remove extra empty line

File: bento/src/main/java/com/yelp/android/bento/core/Component.java [hunk 6]
```diff
@@ -298,6 +306,32 @@ public void setSpanSizeLookup(@NonNull SpanSizeLookup lookup) {
...
 
+    /**
+     * Override this method to handle reordering of items.
+     *
+     * @param oldIndex The index the item was originally in.
+     * @param newIndex The index the item was moved to.
+     * @see Component#isReorderable() (boolean)
+     */
+    public void onItemsMoved(int oldIndex, int newIndex) {
+
```
Comment (dwaxemberg) [inline bento/src/main/java/com/yelp/android/bento/core/Component.java]: Remove empty line

File: bento/src/main/java/com/yelp/android/bento/componentcontrollers/RecyclerViewComponentController.java [hunk 7]
```diff
@@ -30,8 +35,11 @@
 import kotlin.sequences.Sequence;
 import org.jetbrains.annotations.Nullable;
 
-/** Implementation of {@link ComponentController} for {@link RecyclerView}s. */
-public class RecyclerViewComponentController implements ComponentController {
+/**
+ * Implementation of {@link ComponentController} for {@link RecyclerView}s.
+ */
+public class RecyclerViewComponentController implements ComponentController,
+        OnItemMovedPositionListener {
```
Comment (dwaxemberg) [inline bento/src/main/java/com/yelp/android/bento/componentcontrollers/RecyclerViewComponentController.java:42]: What about other ComponentControllers?

File: bento/src/main/java/com/yelp/android/bento/componentcontrollers/RecyclerViewComponentController.java [hunk 8]
```diff
@@ -268,6 +283,27 @@ public void scrollToComponentWithOffset(@NonNull Component component, int offset
         }
     }
 
+    @Override
+    public void onItemMovedPosition(int oldIndex, int newIndex) {
+        RangedValue<Component> rangedValue = mComponentGroup.getLowestRangeValue(oldIndex);
+        rangedValue.mValue.onItemsMoved(oldIndex - rangedValue.mRange.mLower,
+                newIndex - rangedValue.mRange.mLower);
+
+        // bind is not called again, so we need to go through and properly set all the positions
```
Comment (dwaxemberg) [inline bento/src/main/java/com/yelp/android/bento/componentcontrollers/RecyclerViewComponentController.java]: Capital first letter and period at the end.

File: bento/src/main/java/com/yelp/android/bento/componentcontrollers/RecyclerViewComponentController.java [hunk 9]
```diff
@@ -268,6 +283,27 @@ public void scrollToComponentWithOffset(@NonNull Component component, int offset
...
+        int currentIndex = Math.min(oldIndex, newIndex);
+        int highIndex = Math.max(oldIndex, newIndex);
+        while (currentIndex <= highIndex) {
+            ((ViewHolderWrapper) mRecyclerView.findViewHolderForAdapterPosition(currentIndex)).mViewHolder.setAbsolutePosition(currentIndex++);
```
Comment (dwaxemberg) [inline bento/src/main/java/com/yelp/android/bento/componentcontrollers/RecyclerViewComponentController.java]: This line is way too long.

Also, we should just iterate through the visible view holders, all others will have bind called on them when the view scrolls.

File: bento/src/main/java/com/yelp/android/bento/components/OnItemMovedCallback.kt [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+ */
+interface OnItemMovedCallback<T> {
+
+    /**
+     * Called when an item has been dropped somewhere else in the list.
+     * @param oldIndex The old index of the item.
+     * @param newIndex The index where the item landed.
+     * @param newData A new list of data after the reorder.
+     */
+    fun onItemMoved(oldIndex: Int, newIndex: Int, newData: List<T>)
```
Comment (dwaxemberg) [inline bento/src/main/java/com/yelp/android/bento/components/OnItemMovedCallback.kt]: This needs to be used by all components, not just list components therefore we shouldn't make `newData` a list. instead ListComponent should use a `OnItemMovedCallback<List<T>>` in which case `newData` becomes `List<T>`.

But as I mentioned somewhere else, why are we passing the data through this listener?

File: bento/src/main/java/com/yelp/android/bento/core/ListItemTouchCallback.kt [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+    }
+
+    override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {
+        // No swipe support for now.
+    }
+
+    override fun clearView(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder) {
+        super.clearView(recyclerView, viewHolder)
+
+        // This simply checks when the user "drops" an item. onMove is called anytime the item is
```
Comment (dwaxemberg) [inline bento/src/main/java/com/yelp/android/bento/core/ListItemTouchCallback.kt:71]: This says "onMove is called anytime the item..." but it's in the `clearView` method. Can you elaborate a bit?

File: bento/src/main/java/com/yelp/android/bento/core/ListItemTouchCallback.kt [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+
+    private var dragFrom = -1
+    private var dragTo = -1
+
+    override fun canDropOver(
+            recyclerView: RecyclerView,
+            current: RecyclerView.ViewHolder,
+            target: RecyclerView.ViewHolder
+    ): Boolean {
+        // Only allow reorder if it is within the same component.
```
Comment (dwaxemberg) [inline bento/src/main/java/com/yelp/android/bento/core/ListItemTouchCallback.kt:22]: What if I have a ComponentGroup full of Components with 1 view and I want to swap them?
For example, if I wanted to move items inside a CarouselComponent or some sort of grid-like ComponentGroup.

File: bento/src/main/java/com/yelp/android/bento/core/ComponentViewHolder.kt [hunk 13]
```diff
@@ -16,6 +16,12 @@ import android.view.ViewGroup
  */
 abstract class ComponentViewHolder<P, T> {
 
+    /**
+     * Contains the absolute position within the entire [ComponentController]. This should only be
+     * set by the [ComponentController].
+     */
+    var absolutePosition: Int = -1
```
Comment (dwaxemberg) [inline bento/src/main/java/com/yelp/android/bento/core/ComponentViewHolder.kt:24]: Give this an internal setter and public getter to guarantee only the ComponentController (or at least only the library) can set it.

File: bento/src/main/java/com/yelp/android/bento/core/Component.java [hunk 14]
```diff
@@ -184,17 +188,6 @@ public int getNumberLanes() {
...
     }
 
-    /**
-     * Returns the number of lanes at a given position. This will usually return the same as
-     * {@link Component#getNumberLanes()}, but {@link ComponentGroup}s will want to return the
-     * number of lanes at a given index. For example, if you have a {@link ComponentGroup} with
-     * two {@link Component}s with a different number of lanes, this should return the number of
-     * lanes in the component that owns the position.
-     */
-    public int getNumberLanesAtPosition(int position) {
```
Comment (dwaxemberg) [inline bento/src/main/java/com/yelp/android/bento/core/Component.java:194]: Why are we removing this? It seems unrelated to the changes being made.

File: bento/src/main/java/com/yelp/android/bento/core/Component.java [hunk 15]
```diff
@@ -298,6 +306,32 @@ public void setSpanSizeLookup(@NonNull SpanSizeLookup lookup) {
...
+    }
+
+    /**
+     * Returns the lowest component in the component tree. In general, a {@link ComponentGroup} will
+     * recursively call this method, removing the lower bound of its range each time. A
+     * {@link Component} should just return itself.
+     * @param index The index to search for.
+     * @return The lowest component (leaf) in the component tree.
+     */
+    public Component getLowestComponentAtIndex(int index) {
```
Comment (dwaxemberg) [inline bento/src/main/java/com/yelp/android/bento/core/Component.java]: Instead of putting this in the Component class, I think we should do something similar to the `findComponentOffset` in ComponentGroup where we just recursively search the children.

Comment (dwaxemberg) [reply bento/src/main/java/com/yelp/android/bento/core/Component.java]: Also, what if `index` is not `0 <= index < count`

File: bento/src/main/java/com/yelp/android/bento/core/ComponentGroup.java [hunk 16]
```diff
@@ -412,6 +407,41 @@ public int findComponentOffset(@NonNull Component component) {
...
+        Range range = rangeOf(componentAtPos);
+        if (range == null) {
+            throw new IllegalStateException("Found component, but no range associated.");
+        }
+
+        return componentAtPos.getLowestComponentAtIndex(index - range.mLower);
+    }
+
+    /**
+     * Similar to {@link #getLowestComponentAtIndex(int)}, but returns both the component and the
```
Comment (dwaxemberg) [inline bento/src/main/java/com/yelp/android/bento/core/ComponentGroup.java]: If these methods do the same thing, couldn't we have `getLowestComponentAtIndex(int)` return `getLowestRangeValue(int index).mValue`?

File: bento/src/main/java/com/yelp/android/bento/core/ComponentGroup.java [hunk 17]
```diff
@@ -412,6 +407,41 @@ public int findComponentOffset(@NonNull Component component) {
...
+     * absolute range within the controller.
+     *
+     * @param index The index to search for.
+     * @return Both a component and an absolute range over the entire controller.
+     */
+    @Override
+    public RangedValue<Component> getLowestRangeValue(int index) {
```
Comment (dwaxemberg) [inline bento/src/main/java/com/yelp/android/bento/core/ComponentGroup.java]: This method seems very similar to `findComponentOffset` except it includes the upper bound as well.

**Review by dwaxemberg — CHANGES_REQUESTED** [review]

File: bento-sample-app/src/main/java/com/yelp/android/bentosampleapp/ReorderListActivity.kt (hunk 2 above)
Comment (wyattjsmith1) [reply bento-sample-app/src/main/java/com/yelp/android/bentosampleapp/ReorderListActivity.kt:51]: So, `Char.plus()` doesn't concatenate, it adds. Basically, I want to distinguish between each section, so one section is numbers, one is lowercase, and one is uppercase. The `plus` function adds to the Char's ascii value. It is the Java equivalent to `'a' + 1 = 'b'`.

File: bento/src/main/java/com/yelp/android/bento/components/ListComponent.java (hunk 1 above)
Comment (wyattjsmith1) [reply bento/src/main/java/com/yelp/android/bento/components/ListComponent.java]: Ah, this was needed at one time, but not since we added the absolute position to the ViewHolders. I'll remove it.

File: bento/src/main/java/com/yelp/android/bento/componentcontrollers/RecyclerViewComponentController.java (hunk 7 above)
Comment (wyattjsmith1) [reply bento/src/main/java/com/yelp/android/bento/componentcontrollers/RecyclerViewComponentController.java:42]: I was thinking about this. In my opinion, there isn't really a good way to implement this in other controllers. How would a `ViewPager` reorder? Android doesn't really have a built in way to do that.

Additionally, the state of being reorderable is managed by the Component, not the Controller. In that way, Components are really requesting to be reorderable, but it is up to the Controller to reorder. Not sure if this is what we are looking for.

File: bento/src/main/java/com/yelp/android/bento/componentcontrollers/RecyclerViewComponentController.java (hunk 9 above)
Comment (wyattjsmith1) [reply bento/src/main/java/com/yelp/android/bento/componentcontrollers/RecyclerViewComponentController.java]: Hmm... ok. How would we iterate only the visible view holders?

File: bento/src/main/java/com/yelp/android/bento/components/OnItemMovedCallback.kt (hunk 10 above)
Comment (wyattjsmith1) [reply bento/src/main/java/com/yelp/android/bento/components/OnItemMovedCallback.kt]: Removed in comment above.

File: bento/src/main/java/com/yelp/android/bento/core/Component.java (hunk 14 above)
Comment (wyattjsmith1) [reply bento/src/main/java/com/yelp/android/bento/core/Component.java:194]: Eh, a little unrelated, but still related. I find that both grids and reordering require me to be able to find a component at a given absolute index. I just created a getLowestComponentAtIndex, then call whatever method from there. Calling `getLowestComponentAtIndex(n).getNumberLanes()` seems cleaner than having a bunch of redundant methods that recursively find the lowest component, then return some field from it.

File: bento/src/main/java/com/yelp/android/bento/core/ComponentViewHolder.kt (hunk 13 above)
Comment (wyattjsmith1) [reply bento/src/main/java/com/yelp/android/bento/core/ComponentViewHolder.kt:24]: The problem is that doesn't compile. I think it is because the RecyclerViewComponentController is in java, which doesn't recognize internal, and it is probably package protected then.

File: bento/src/main/java/com/yelp/android/bento/core/ComponentGroup.java (hunk 17 above)
Comment (wyattjsmith1) [reply bento/src/main/java/com/yelp/android/bento/core/ComponentGroup.java]: You aren't wrong. `findComponentOffset` takes a component and  `getLowestRangeValue` takes an index. Are you suggesting something?

File: bento/src/main/java/com/yelp/android/bento/core/ListItemTouchCallback.kt (hunk 12 above)
Comment (wyattjsmith1) [reply bento/src/main/java/com/yelp/android/bento/core/ListItemTouchCallback.kt:22]: I agree that this implementation is limiting. I guess my questions is, do you have any other suggestions? I suppose we could have some default implementation of this in `Component` that does this check, and override it in `ComponentGroup` if needed. Do you think that would be better?

File: bento-sample-app/src/main/java/com/yelp/android/bentosampleapp/ReorderListActivity.kt [hunk 18]
```diff
@@ -0,0 +1,107 @@
...
+
+class ReorderViewHolder : ComponentViewHolder<Presenter, String>() {
+
+    private lateinit var text: TextView
+    private lateinit var presenter: Presenter
+
+    override fun inflate(parent: ViewGroup): View {
+        return parent.inflate<View>(R.layout.reorderable_view_holder).also {
+                    text = it.findViewById(R.id.textview)
+                    it.findViewById<View>(R.id.handle).setOnTouchListener { _, event ->
```
Comment (wyattjsmith1) [inline bento-sample-app/src/main/java/com/yelp/android/bentosampleapp/ReorderListActivity.kt]: @dwaxemberg Do you think it would be useful to add some sort of extension function such as `setOnDragStartListener()` that would take a lambda? It feels weird to add a touch listener and listen for `ACTION_DOWN`. It could be used like so:

```
view.setOnDragStartListener {
  presenter.onStartDrag(this)
}

File: bento/src/main/java/com/yelp/android/bento/core/ListItemTouchCallback.kt (hunk 12 above)
Comment (wyattjsmith1) [reply bento/src/main/java/com/yelp/android/bento/core/ListItemTouchCallback.kt:22]: Another thing to remember is that the `RecyclerView`'s reordering only works with `ViewHolder`s. If we were to allow entire components to be reordered, we would have to check that their count is only 1 or some weird things might happen.

Comment (wyattjsmith1) [reply bento/src/main/java/com/yelp/android/bento/core/ListItemTouchCallback.kt:22]: Another thing I'm gonna throw out is that we may not want to allow everything within one component be reorderable. Consider this layout. It is a ListComponent, where we add another item to the end to view/add photos. We probably don't want to be able to rearrange a photo behind the button.
image

File: bento/src/main/java/com/yelp/android/bento/componentcontrollers/RecyclerViewComponentController.java (hunk 9 above)
Comment (dwaxemberg) [reply bento/src/main/java/com/yelp/android/bento/componentcontrollers/RecyclerViewComponentController.java]: You can ask the layout manager for the first and last visible indices. We do a similar thing in the VisibilityListener.

File: bento/src/main/java/com/yelp/android/bento/core/ListItemTouchCallback.kt (hunk 12 above)
Comment (dwaxemberg) [reply bento/src/main/java/com/yelp/android/bento/core/ListItemTouchCallback.kt:22]: Yeah I can see how this would get complicated, maybe it should be a follow up. I do think we should support reordering of components within a component group though. We would definitely need to figure out a way to guarantee these are single item components to avoid any messy RecyclerView issues. 

I guess for now just forget about it and maybe open an issue as "nice to have".

File: bento/src/main/java/com/yelp/android/bento/componentcontrollers/RecyclerViewComponentController.java (hunk 7 above)
Comment (dwaxemberg) [reply bento/src/main/java/com/yelp/android/bento/componentcontrollers/RecyclerViewComponentController.java:42]: I was thinking of a ListView primarily, but that might also be difficult to handle. Maybe we should just stick to RecylerViews for now.

File: bento/src/main/java/com/yelp/android/bento/core/Component.java (hunk 14 above)
Comment (dwaxemberg) [reply bento/src/main/java/com/yelp/android/bento/core/Component.java:194]: Gotcha. I don't like the term "lowest" to refer to the leaf node however. Also, since this is not just returning a field, I think we should change to "find" instead of "get" so it doesn't give the impression that this is a costless operation. What do you think about renaming it to something like `findIndexOwner` or `findComponentWithIndex`?

File: bento/src/main/java/com/yelp/android/bento/core/ComponentGroup.java (hunk 17 above)
Comment (dwaxemberg) [reply bento/src/main/java/com/yelp/android/bento/core/ComponentGroup.java]: Could we combine the two methods somehow? (e.g. make a helper method that both methods call into in order to reduce some of the very similar logic)

Comment (dwaxemberg) [reply bento/src/main/java/com/yelp/android/bento/core/ComponentGroup.java]: Or maybe just give them complimentary names (like `findComponentWithIndex` and `findIndexOfComponent` or something)

File: bento-sample-app/src/main/java/com/yelp/android/bentosampleapp/ReorderListActivity.kt (hunk 18 above)
Comment (dwaxemberg) [reply bento-sample-app/src/main/java/com/yelp/android/bentosampleapp/ReorderListActivity.kt]: Generally I like the idea, but would it be useful to have it in the library vs. making the consumers declare that function if they want it?

Comment (wyattjsmith1) [reply bento-sample-app/src/main/java/com/yelp/android/bentosampleapp/ReorderListActivity.kt]: Spoke on Slack, we agreed it makes sense in the Bento library.

File: bento/src/main/java/com/yelp/android/bento/core/ListItemTouchCallback.kt (hunk 12 above)
Comment (wyattjsmith1) [reply bento/src/main/java/com/yelp/android/bento/core/ListItemTouchCallback.kt:22]: Whoops, looks like Github lost my comment. TBH, I  think this would be difficult to patch in later. Additionally, the current implementation would not work with the example above, which is the primary reason I am implementing this.

One possible solution is to create a function `canDropItem(fromComponent, fromIndex, toComponent, toIndex)` in each `Component` that allows us to control whether a `Component` can take a given item. 

This solution would cover the case where a component doesn't want to allow an item to be dropped in a certain location (like my example above), but does not consider a `ComponentGroup` with `Component`s with a count > 1. It seems like the only way a `ComponentGroup` could be reorderable is if all of its children are `Component`s with a count == 1. I suppose we could do this check, but it wouldn't be the greatest. Additionally, what would we do when a child has a count > 1? Throw an exception? Not reorder? Just roll with it and hope for the best?

Comment (wyattjsmith1) [reply bento/src/main/java/com/yelp/android/bento/core/ListItemTouchCallback.kt:22]: Ok, I added a solution for this. The only downside I see if that we now have `findRangedComponentWithIndex` and `findReorderTargetAtIndex` which have very similar implementations, but I don't see a way to combine them.

File: bento/src/main/java/com/yelp/android/bento/core/Component.java [hunk 19]
```diff
@@ -298,6 +303,27 @@ public void setSpanSizeLookup(@NonNull SpanSizeLookup lookup) {
...
+     *
+     * @param oldIndex The index the item was originally in.
+     * @param newIndex The index the item was moved to.
+     * @see Component#isReorderable() (boolean)
+     */
+    public void onItemsMoved(int oldIndex, int newIndex) {
+    }
+
+    /**
+     * Checks if a n item from one component can be dropped in this component at a given index.
```
Comment (dwaxemberg) [inline bento/src/main/java/com/yelp/android/bento/core/Component.java]: typo: remove the space before the 'n' in 'an'

File: bento/src/main/java/com/yelp/android/bento/core/ComponentGroup.java [hunk 20]
```diff
@@ -412,6 +407,58 @@ public int findComponentOffset(@NonNull Component component) {
...
+     * Finds and returns the component at lowest level (leaf) that encompasses the index.
+     *
+     * @param index The index to search for.
+     * @return The lowest component in the tree.
+     */
+    public Component findComponentWithIndex(int index) {
+        return findRangedComponentWithIndex(index).mValue;
+    }
+
+    public RangedValue<Component> findReorderTargetAtIndex(int index) {
```
Comment (dwaxemberg) [inline bento/src/main/java/com/yelp/android/bento/core/ComponentGroup.java]: Add a javadoc for this

File: bento/src/main/java/com/yelp/android/bento/core/ListItemTouchCallback.kt [hunk 21]
```diff
@@ -0,0 +1,101 @@
...
+
+    override fun getMovementFlags(
+            recyclerView: RecyclerView,
+            viewHolder: RecyclerView.ViewHolder
+    ): Int {
+        val lowestComponent = component.findReorderTargetAtIndex(viewHolder.adapterPosition).mValue
+        if (lowestComponent.isReorderable) {
+            if (lowestComponent is ComponentGroup) {
+                for (index in (0 until lowestComponent.size)) {
+                    if (lowestComponent[index].count != 1) {
```
Comment (dwaxemberg) [inline bento/src/main/java/com/yelp/android/bento/core/ListItemTouchCallback.kt]: Can we use `asSequence` here?
```
if (lowestComponent.asSequence().any { it.count != 1 }) {
...
```

Also, I think we should check >0 instead of != 1 because if we have a bunch of components with a count of 0 we should be able to safely ignore them.

Comment (wyattjsmith1) [reply bento/src/main/java/com/yelp/android/bento/core/ListItemTouchCallback.kt]: I assume you mean > 1?

File: bento/src/main/java/com/yelp/android/bento/componentcontrollers/ListViewComponentController.kt [hunk 22]
```diff
@@ -15,6 +15,7 @@ import com.yelp.android.bento.core.ComponentController
 import com.yelp.android.bento.core.ComponentGroup
 import com.yelp.android.bento.core.ComponentViewHolder
 import com.yelp.android.bento.core.ComponentVisibilityListener
+import com.yelp.android.bento.core.GapViewHolder
```
Comment (wyattjsmith1) [inline bento/src/main/java/com/yelp/android/bento/componentcontrollers/ListViewComponentController.kt]: Whoops, removing.

File: bento/src/main/java/com/yelp/android/bento/components/ListComponent.java [hunk 23]
```diff
@@ -253,6 +255,38 @@ public int getSpanSize(int position) {
         };
     }
 
+    @Override
+    public void onItemsMoved(
```
Comment (dwaxemberg) [inline bento/src/main/java/com/yelp/android/bento/components/ListComponent.java]: Should we make this `final` here?

File: bento/src/main/java/com/yelp/android/bento/components/OnItemMovedCallback.kt [hunk 24]
```diff
@@ -0,0 +1,14 @@
+package com.yelp.android.bento.components
+
+/**
+ * Interface for listening to drag and drop events.
+ */
+interface OnItemMovedCallback<T> {
+
+    /**
+     * Called when an item has been dropped somewhere else in the list.
```
Comment (dwaxemberg) [inline bento/src/main/java/com/yelp/android/bento/components/OnItemMovedCallback.kt]: nit: change "list" to "component"

File: bento/src/main/java/com/yelp/android/bento/core/ComponentGroup.java [hunk 25]
```diff
@@ -40,6 +44,9 @@ public ComponentGroup() {
         mSpanSizeLookup = new SpanSizeLookup() {
             @Override
             public int getSpanSize(int position) {
+                if (hasGap(position)) {
```
Comment (dwaxemberg) [inline bento/src/main/java/com/yelp/android/bento/core/ComponentGroup.java]: This looks like a bug fix for something. Is it related to this PR or should we make it separate?

File: bento/src/main/java/com/yelp/android/bento/core/ComponentGroup.java [hunk 26]
```diff
@@ -412,6 +410,43 @@ public int findComponentOffset(@NonNull Component component) {
...
+    }
+
+    /**
+     * Returns both the component and the absolute range within the controller.
+     *
+     * @param index The index to search for.
+     * @return Both a component and an absolute range over the entire controller.
+     */
+    public RangedValue<Component> findRangedComponentWithIndex(int index) {
+
```
Comment (dwaxemberg) [inline bento/src/main/java/com/yelp/android/bento/core/ComponentGroup.java]: remove extra line

File: bento/src/main/java/com/yelp/android/bento/components/ListComponent.java (hunk 23 above)
Comment (wyattjsmith1) [reply bento/src/main/java/com/yelp/android/bento/components/ListComponent.java]: Eh, I can't imagine a reason someone would want to override this.

File: bento/src/main/java/com/yelp/android/bento/core/ComponentGroup.java (hunk 25 above)
Comment (wyattjsmith1) [reply bento/src/main/java/com/yelp/android/bento/core/ComponentGroup.java]: Eh, back when we were testing `ComponentGroups` with gaps, it crashed and this fixed it. Now that we eliminated that feature, the test app works. I will open a separate PR for this.

**Review by dwaxemberg — APPROVED** [review]

**Comment by dwaxemberg** [comment]
Don't forget to bump the version
